src/frontegg-auth/src/auth.rs

```diff
@@ -7,82 +7,269 @@
+
+        Self {
+            validation_config,
+            metrics,
+            admin_api_token_url: config.admin_api_token_url,
             client,
+            inflight_auth_requests,
         }
     }
 
+    /// Create an [`Authentication`] from [`FronteggCliArgs`].
+    pub fn from_args(
+        args: FronteggCliArgs,
+        registry: &MetricsRegistry,
+    ) -> Result<Option<Self>, Error> {
+        let config = match (
+            args.frontegg_tenant,
+            args.frontegg_api_token_url,
+            args.frontegg_jwk,
+            args.frontegg_admin_role,
+        ) {
+            (None, None, None, None) => {
+                return Ok(None);
+            }
+            (Some(tenant_id), Some(admin_api_token_url), Some(jwk), Some(admin_role)) => {
+                AuthenticationConfig {
+                    admin_api_token_url,
+                    decoding_key: DecodingKey::from_rsa_pem(jwk.as_bytes())?,
+                    tenant_id: Some(tenant_id),
+                    now: mz_ore::now::SYSTEM_TIME.clone(),
+                    admin_role,
+                }
+            }
+            _ => unreachable!("clap enforced"),
+        };
+        let client = Client::environmentd_default();
+
+        Ok(Some(Self::new(config, client, registry)))
+    }
+
     /// Exchanges a password for an access token and a refresh token.
     pub async fn exchange_password_for_token(
         &self,
         password: &str,
-    ) -> Result<ApiTokenResponse, Error> {
+        expected_email: String,
+    ) -> Result<ExchangePasswordForTokenResponse, Error> {
         let password: AppPassword = password.parse()?;
-        self.client
-            .exchange_client_secret_for_token(
-                password.client_id,
-                password.secret_key,
-                &self.admin_api_token_url,
-            )
-            .await
+        let req = ApiTokenArgs {
+            client_id: password.client_id,
+            secret: password.secret_key,
+        };
+
+        // Note: we get the reciever in a block to scope the access to the mutex.
+        let rx = {
+            let mut inflight = self
+                .inflight_auth_requests
+                .lock()
+                .expect("Frontegg Auth Client panicked");
+            let (tx, rx) = tokio::sync::oneshot::channel();
+
+            match inflight.get_mut(&req) {
+                // Already have an inflight request, add to our list of waiters.
+                Some(senders) => {
+                    tracing::debug!(?req, "reusing request");
+                    senders.push(tx);
+                    rx
+                }
+                // New request! Need to queue one up.
+                None => {
+                    tracing::debug!(?req, "spawning new request");
+
+                    inflight.insert(req.clone(), vec![tx]);
+
+                    let client = self.client.clone();
+                    let inflight = Arc::clone(&self.inflight_auth_requests);
```

Comment:
i think this implicitly drops the previous lock guard...can you make that explicit?

Reply:
Good call, I made this an explicit drop